src/Spice86.Core/Emulator/InterruptHandlers/Input/Keyboard/BiosKeyboardInt9Handler.cs

```diff
@@ -6,35 +6,96 @@
 using Spice86.Core.Emulator.Devices.ExternalInput;
 using Spice86.Core.Emulator.Devices.Input.Keyboard;
 using Spice86.Core.Emulator.Function;
+using Spice86.Core.Emulator.InterruptHandlers.Common.MemoryWriter;
 using Spice86.Core.Emulator.Memory;
+using Spice86.Shared.Emulator.Memory;
 using Spice86.Shared.Interfaces;
 
 /// <summary>
-/// Crude implementation of BIOS keyboard buffer handler (hardware interrupt 0x9, IRQ1)
+/// Implementation of BIOS keyboard buffer handler (hardware interrupt 0x9, IRQ1)
 /// </summary>
 public class BiosKeyboardInt9Handler : InterruptHandler {
-    private readonly Keyboard _keyboard;
+    private readonly AvaloniaKeyConverter _scanCodeConverter = new();
+    private readonly PS2Keyboard _keyboard;
+    private static readonly SegmentedAddress CallbackLocation = new(0xf000, 0xe987);
     private readonly DualPic _dualPic;
 
     /// <summary>
     /// Initializes a new instance.
     /// </summary>
     /// <param name="memory">The memory bus.</param>
-    /// <param name="functionHandlerProvider">Provides current call flow handler to peek call stack.</param>
     /// <param name="stack">The CPU stack.</param>
     /// <param name="state">The CPU state.</param>
-    /// <param name="dualPic">The two programmable interrupt controllers.</param>
-    /// <param name="keyboard">The keyboard controller.</param>
+    /// <param name="functionHandlerProvider">Provides current call flow handler to peek call stack.</param>
+    /// <param name="keyboard">The keyboard device for direct port access.</param>
     /// <param name="biosKeyboardBuffer">The structure in emulated memory this interrupt handler writes to.</param>
     /// <param name="loggerService">The logger service implementation.</param>
-    public BiosKeyboardInt9Handler(IMemory memory,
-        IFunctionHandlerProvider functionHandlerProvider, Stack stack,
-        State state, DualPic dualPic, Keyboard keyboard,
-        BiosKeyboardBuffer biosKeyboardBuffer, ILoggerService loggerService)
+    public BiosKeyboardInt9Handler(IMemory memory, Stack stack, State state,
+        IFunctionHandlerProvider functionHandlerProvider, DualPic dualPic,
+        PS2Keyboard keyboard, BiosKeyboardBuffer biosKeyboardBuffer,
+        ILoggerService loggerService)
         : base(memory, functionHandlerProvider, stack, state, loggerService) {
+        BiosKeyboardBuffer = biosKeyboardBuffer;
         _keyboard = keyboard;
         _dualPic = dualPic;
-        BiosKeyboardBuffer = biosKeyboardBuffer;
+    }
+
+    public override SegmentedAddress WriteAssemblyInRam(MemoryAsmWriter memoryAsmWriter) {
+        SegmentedAddress savedAddress = memoryAsmWriter.CurrentAddress;
+        memoryAsmWriter.CurrentAddress = CallbackLocation;
+
+        // push ax (save original AX)
+        memoryAsmWriter.WriteUInt8(0x50);
```

Comment:
Why cant this be c#?

Reply:
Fewer dependencies. I don't have to depend on EmulationLoopRecall.

Plus it would be more code in C# and some parts like EOI stuff were easier this way.

TSR are also supposed to be able to react to this, this is easier if it's asm.